### PR TITLE
Revendored Swarmkit 

### LINKED ISSUE #34825: Swarm 17.07.0 unable to join nodes
**Description**

Hi, 

Upgrade from docker-ce 17.06.2 to docker-ce-17.07.0 break my cluster.
My manager is up and running but worker (or manager) nodes cannot join it.

**Steps to reproduce the issue:**
1. install docker-ce-17.07.0 on 2 VMs (say M and W)
2. run `docker swarm init --advertise-addr=<M_public_ip>:2377` on M (get the worker join command)
3. run the join command on W

**Describe the results you received:**

`Error response from daemon: rpc error: code = Unavailable desc = grpc: the connection is unavailable`

**Describe the results you expected:**

`This node joined a swarm as a worker.`

**Additional information you deem important (e.g. issue happens only occasionally):**

I used a Terrafom+Ansible that was successfully ran many time to create clusters. 
It even appears on an existing cluster after upgrading to 17.07.0

**Output of `docker version`:**

```
Client:
 Version:      17.07.0-ce
 API version:  1.31
 Go version:   go1.8.3
 Git commit:   8784753
 Built:        Tue Aug 29 17:42:01 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.07.0-ce
 API version:  1.31 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   8784753
 Built:        Tue Aug 29 17:43:23 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 17.07.0-ce
Storage Driver: overlay
 Backing Filesystem: xfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: eqt38cc9zdy5i15yflndsj0ze
 Is Manager: true
 ClusterID: u7yvqf60ew3r1n867bjstz2bh
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Root Rotation In Progress: false
 Node Address: 10.90.251.200
 Manager Addresses:
  10.90.251.200:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 3addd840653146c90a254301d6c3a663c7fd6429
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-514.21.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.797GiB
Name: swarm-mode-latest-master-0
ID: W4TM:QHM2:MWD4:52SK:XMIM:TJA5:2NCX:4WAB:77C5:TF66:WVKX:XGTD
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://some-proxy:3128
Https Proxy: http://some-proxy:3128
No Proxy: localhost,172.17.42.1,.sock
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

The swarm mode cluster run on KVM VMs

### LINKED ISSUE #34996: Worker cannot be joined to swarm cluster
**Description**
Worker cannot be joined to swarm manager.

**Steps to reproduce the issue:**
1. Create manager node:
```
docker swarm init
```
2. Create worker node:
```
docker swarm join --token SWMTKN-1-1m7h3jnhycq46r5qzunx8gwvjjr4f30ckfqet2uqgdj60x3a9a-7kr5m1c3syidtmpgouh54f0fo <ip>:2377
```

**Describe the results you received:**
```
Sep 27 13:07:18 q4de3csye43 dockerd[10836]: time="2017-09-27T13:07:18.304422384+02:00" level=error msg="failed to retrieve remote root CA certificate" error="rpc error: code = Unavailable desc = grpc: the connection is unavailable" module=node
Sep 27 13:07:20 q4de3csye43 dockerd[10836]: time="2017-09-27T13:07:20.307296581+02:00" level=error msg="failed to retrieve remote root CA certificate" error="rpc error: code = Unavailable desc = grpc: the connection is unavailable" module=node
Sep 27 13:07:22 q4de3csye43 dockerd[10836]: time="2017-09-27T13:07:22.309839445+02:00" level=error msg="failed to retrieve remote root CA certificate" error="rpc error: code = Unavailable desc = grpc: the connection is unavailable" module=node
Sep 27 13:07:24 q4de3csye43 dockerd[10836]: time="2017-09-27T13:07:24.312388683+02:00" level=error msg="failed to retrieve remote root CA certificate" error="rpc error: code = Unavailable desc = grpc: the connection is unavailable" module=node
Sep 27 13:07:26 q4de3csye43 dockerd[10836]: time="2017-09-27T13:07:26.315234191+02:00" level=error msg="failed to retrieve remote root CA certificate" error="rpc error: code = Unavailable desc = grpc: the connection is unavailable" module=node
Sep 27 13:07:28 q4de3csye43 dockerd[10836]: time="2017-09-27T13:07:28.315699672+02:00" level=error msg="cluster exited with error: rpc error: code = Unavailable desc = grpc: the connection is unavailable"
Sep 27 13:07:28 q4de3csye43 dockerd[10836]: time="2017-09-27T13:07:28.316059555+02:00" level=error msg="Handler for POST /v1.32/swarm/join returned error: rpc error: code = Unavailable desc = grpc: the connection is unavailable"
```

**Describe the results you expected:**
Worker is joined.

**Additional information you deem important (e.g. issue happens only occasionally):**
We faced to this issue after Docker was upgraded from 17.06.2-ce to 17.09.0-ce. Now it is not possible to join any worker to the swarm.

**Output of `docker version`:**

```
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:41:23 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:42:49 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 15
Server Version: 17.09.0-ce
Storage Driver: overlay
 Backing Filesystem: extfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: keosiajh7onqpn42u3ixsrmb3
 Is Manager: true
 ClusterID: qe8cswquhtjl817eiu4gtzyio
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 10.33.177.47
 Manager Addresses:
  <ip>:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-514.21.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 9.601GiB
Name: q4de3csye47
ID: 2FXY:XP3H:UWFL:3R7U:CF5H:THXJ:XC5K:SZTS:U4LV:AY7D:3JNV:FOJF
Docker Root Dir: /docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://<ip>:3128/
Https Proxy: http://<ip>:3128/
No Proxy: <ip>,registry
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 10.33.177.26:5000
 registry:5000
 127.0.0.0/8
Live Restore Enabled: false

WARNING: bridge-nf-call-ip6tables is disabled
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
OS: Centos 7
Linux q4de3csye43 3.10.0-693.2.2.el7.x86_64 #1 SMP Tue Sep 12 22:26:13 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

### LINKED ISSUE #35046: 17.07-ce Docker Swarm - Can't join another manager or promote a worker (initialize raft node: rpc error)
When I try to add a node as a second manager in swarm
````
docker swarm join --listen-addr THIS_NODE_IP:2377 --advertise-addr THIS_NODE_IP --token "MANAGER_TOKEN" MANAGER-1_NODE_IP:2377
````
or promote an existing worker
````
(on a worker "master-2") docker swarm join --listen-addr THIS_NODE_IP:2377 --advertise-addr THIS_NODE_IP --token "WORKER_TOKEN" MANAGER-1_NODE_IP:2377
>>> node has successfully joined the swarm

(on swarm manager & leader) docker node promote master-2
````

I am getting errors `can't initialize raft node: rpc error: code = Unknown desc` and node becomes DOWN for Swarm Leader.

Setup:

Docker 1.17.07 (Swarm Mode), RHEL 7.3, Kernel 4.13.3 (same behavior on 3.10), behind proxy.

**Expand below items for docker info output:**
<details><summary>docker info (swarm master-1)</summary>

````
root@master-1 # docker info
Containers: 3
 Running: 3
 Paused: 0
 Stopped: 0
Images: 4
Server Version: 17.07.0-ce
Storage Driver: overlay
 Backing Filesystem: extfs
 Supports d_type: true
Logging Driver: journald
Cgroup Driver: cgroupfs
Plugins:
 Volume: local (*using volume driver=local, type=nfs*)
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: aw5r20gz2pr18m9yglk4qwjh4
 Is Manager: true
 ClusterID: b5gds92q3otj9ego2en95qm34
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Root Rotation In Progress: false
 Node Address: <manager-1-ip>
 Manager Addresses:
<manager-1-ip>:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 3addd840653146c90a254301d6c3a663c7fd6429
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.13.3-1.el7.elrepo.x86_64
Operating System: Red Hat Enterprise Linux Server 7.3 (Maipo)
OSType: linux
Architecture: x86_64
Name: master-1
ID: DMJT:S6EF:LUAF:PQKX:I5R6:TU2W:JZRH:MVMM:YJC5:MC2K:SGSQ:PXKM
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://<proxy>
Https Proxy: http://<proxy>
No Proxy: <nodes-ips>,<master-1-ip>,<master-2-ip>,<proxy-ip>,127.0.0.1,localhost,sonar,jenkins
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
````
</details>
<details><summary>docker info (swarm master-2)</summary>

````
root@master-2 # docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 17.07.0-ce
Storage Driver: overlay
 Backing Filesystem: xfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 3addd840653146c90a254301d6c3a663c7fd6429
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-514.el7.x86_64
Operating System: Red Hat Enterprise Linux Server 7.3 (Maipo)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 8GiB
Name: master-2
ID: EY7E:5F7I:OEDA:PBXW:ETM2:IQZU:USYK:6UQM:NGVT:HEFU:TJ3Q:QAUP
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://<proxy>
Https Proxy: http://<proxy>
No Proxy: <nodes-ips>,<master-1-ip>,<master-2-ip>,<proxy-ip>,127.0.0.1,localhost,sonar,jenkins
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
````
</details><details><summary>docker info (swarm node)</summary>

````
Containers: 5
 Running: 5
 Paused: 0
 Stopped: 0
Images: 8
Server Version: 17.07.0-ce
Storage Driver: overlay
 Backing Filesystem: extfs
 Supports d_type: true
Logging Driver: journald
Cgroup Driver: cgroupfs
Plugins:
 Volume: local (*use volume driver=local type=nfs*)
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: ph3qqr9ti750cc3m741rs1dsm
 Is Manager: false
 Node Address:  <node-ip>
 Manager Addresses:
  <manager-ip>:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 3addd840653146c90a254301d6c3a663c7fd6429
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 seccomp 
  Profile: default
Kernel Version: 4.13.3-1.el7.elrepo.x86_64
Operating System: Red Hat Enterprise Linux Server 7.3 (Maipo)
OSType: linux
Architecture: x86_64
Name: node1
ID: KR3X:353P:3XHC:Z7M3:ZXU3:W4NI:7N4O:VYMI:MTYR:D47M:UMT4:LKH5
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://<proxy>
Https Proxy: http://<proxy>
No Proxy: <nodes-ips>,<master-1-ip>,<master-2-ip>,<proxy-ip>,127.0.0.1,localhost,sonar,jenkins
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
````
</details>
---

Steps to reproduce:
1. on master-1: `docker swarm init`
2. on master-1: `docker swarm join-token worker` 
    on node-1:   `docker swarm join --listen-addr THIS_NODE_IP:2377 --advertise-addr THIS_NODE_IP --token "WORKER_TOKEN" MANAGER-1_NODE_IP:2377` 
3. on master-1:  `docker swarm join-token manager`
    on master-2: `docker swarm join --listen-addr THIS_NODE_IP:2377 --advertise-addr THIS_NODE_IP --token "MANAGER_TOKEN" MANAGER-1_NODE_IP:2377`
4. "alternative for 3 - join as worker, then promote":
   on master-1:  `docker swarm join-token worker`
   on master-2: `docker swarm join --listen-addr THIS_NODE_IP:2377 --advertise-addr THIS_NODE_IP --token "WORKER_TOKEN" MANAGER-1_NODE_IP:2377`
   on master-1:  `docker swarm promote master-2`

**Expected Result:**
Node should successfully join Swarm as Manager and on Leader node `docker node ls` should display this node as Reachable.

**Actual Result**
Either joining via manager token or "worker"+"promote" method, node fails with the message `cluster exited with error: manager stopped: can't initialize raft node: rpc error: code = Unknown desc = could not connect to prospective new cluster member using its advertised address: rpc error: code = Unavailable desc = grpc: the connection is unavailable`,

`docker node ls` on the Leader shows a manager candidate node status DOWN.
logs:
````
Sep 30 19:10:46 master-2 dockerd: time="2017-09-30T19:10:46.400925292+02:00" level=info msg="Stopping manager" module=node node.id=7r92yw3pcfcjy4f299dwfwy4l
Sep 30 19:10:46 master-2 dockerd: time="2017-09-30T19:10:46.401013876+02:00" level=info msg="Manager shut down" module=node node.id=7r92yw3pcfcjy4f299dwfwy4l
Sep 30 19:10:46 master-2 dockerd: time="2017-09-30T19:10:46.401086367+02:00" level=info msg="shutting down certificate renewal routine" module="node/tls" node.id=7r92yw3pcfcjy4f299dwfwy4l node.role=swarm-manager
Sep 30 19:10:46 master-2 dockerd: time="2017-09-30T19:10:46.401551984+02:00" level=error msg="cluster exited with error: manager stopped: can't initialize raft node: rpc error: code = Unknown desc = could not connect to prospective new cluster member using its advertised address: rpc error: code = Unavailable desc = grpc: the connection is unavailable"
Sep 30 19:10:46 master-2 dockerd: time="2017-09-30T19:10:46.401601032+02:00" level=warning msg="Restarting swarm in 0.20 seconds"
````

Network connectivity checked (otherwise rest of the swarm wouldn't work). 

Thank you in advance.

### LINKED ISSUE #35395: Swarm worker cannot connect to master if proxy is configured
**Description**

Beginning with Docker CE 17.09, a Swarm worker connects to its master through a configured proxy. Unfortunately this fails because Docker stores the master IP addresses instead of hostnames so the master doesn't match the values configured in `NO_PROXY`.

**Steps to reproduce the issue:**
1. Setup Docker as usual. Add a proxy configuration like this to `/etc/systemd/system/docker.service.d/proxy.conf`:
    ```
    [Service]
    Environment="HTTP_PROXY=http://proxy.internal.company.com:8080" "HTTPS_PROXY=http://proxy.internal.company.com:8080" "NO_PROXY=localhost,127.0.0.1,.internal.company.com"
    ```
    Don't forget to `systemctl daemon-reload` and restart the Docker daemon.

2. Initialize swarm on the master node
    ```
    master# docker swarm init
    Swarm initialized: current node (0tf92vjqqivr9mujdd99ip6ob) is now a manager.

    To add a worker to this swarm, run the following command:

        docker swarm join --token SWMTKN-1-20n86wgp1jlfdku52x5pzlcwc6youavmzm78junqe7wtxqg4t5-2orbo6etiiknwjcfj8g52chtr 192.168.1.11:2377

    To add a manager to this swarm, run 'docker swarm join-token manager' and follow the instructions.
    ```

3. Add a worker as indicated:
    ```
    worker# docker swarm join --token SWMTKN-1-20n86wgp1jlfdku52x5pzlcwc6youavmzm78junqe7wtxqg4t5-2orbo6etiiknwjcfj8g52chtr 192.168.1.11:2377
    ```

    Swarm join hangs because docker attempts to connect to the master through the proxy. Then it reports

    ```
    Error response from daemon: rpc error: code = Unavailable desc = grpc: the connection is unavailable
    ```

    Repeat with a FQDN and the join succeeds right away:

    ```
    # docker swarm join --token SWMTKN-1-20n86wgp1jlfdku52x5pzlcwc6youavmzm78junqe7wtxqg4t5-2orbo6etiiknwjcfj8g52chtr master.internal.company.com:2377
    This node joined a swarm as a worker.
    ```

    However, after a `dockerd` restart the worker canot connect to a master again.

**Describe the results you received:**
See above.

**Describe the results you expected:**
A swarm worker should either ignore the proxy settings, store the master FQDN instead of IP address, or have an option whether to use the proxy for master connections.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
# docker version
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:41:23 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:42:49 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 17.09.0-ce
Storage Driver: zfs
 Zpool: error while getting pool information strconv.ParseUint: parsing "": invalid syntax
 Zpool Health: not available
 Parent Dataset: apppool/docker
 Space Used By Parent: 405504
 Space Available: 20671299584
 Parent Quota: no
 Compression: lz4
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: 6c839q9vfo18ucl8g7yykwukg
 Is Manager: false
 Node Address: 192.168.1.12
 Manager Addresses:
  192.168.1.11:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-693.5.2.el7.x86_64
Operating System: Red Hat Enterprise Linux Server 7.4 (Maipo)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 39.12GiB
Name: worker
ID: PLND:NJOE:YUR6:XJLP:HYEM:IO4I:KS4G:5BYP:ZOOD:BEJH:G2FR:4VVE
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://proxy.internal.company.com:8080
Https Proxy: http://proxy.internal.company.com:8080
No Proxy: localhost,127.0.0.1,.internal.company.com
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
OS is RHEL 7.4 running on vSphere 6.5

### PR DESCRIPTION
**Pull request by RenaudWasTaken**
includes https://github.com/docker/swarmkit/pull/2419, which:

fixes https://github.com/moby/moby/issues/34825
fixes https://github.com/moby/moby/issues/34996
fixes https://github.com/moby/moby/issues/35046
fixes https://github.com/moby/moby/issues/35395

Hello!

**- What I did**
- Revendored swarmkit to the latest commit to include the recent GenericResource change
- Updated the Generic resource CLI to match it's new CLI format
- Fixed the DockerSwarmSuite.TestSwarmVolumePlugin as the error message is now surfaced in a new field

**- Description for the changelog**
Revendored Swarmkit

@anshulpundir can you confirm that the revendoring does not require me to change moby/moby ?
Ping @thaJeztah

**- Bumps**

https://github.com/docker/swarmkit/compare/28f91d87bd3f75fd039dbb9be49bfd2381019261...de950a7ed842c7b7e47e9451cde9bf8f96031894

- Delete node attachments when node is removed https://github.com/docker/swarmkit/pull/2409
- add pid limits https://github.com/docker/swarmkit/pull/2415 (related to https://github.com/moby/moby/issues/28618)
- Provide custom gRPC dialer to override default proxy dialer https://github.com/docker/swarmkit/pull/2419
- Generic resource cli https://github.com/docker/swarmkit/pull/2347

**- A picture of a cute animal (not mandatory but encouraged)**
jack-russell-2-1024x683

### CODE REVIEW

**Comment by yongtang** [comment]
Looks like the following failure happens on all platform:
```
14:55:19 FAIL: docker_cli_swarm_unix_test.go:15: DockerSwarmSuite.TestSwarmVolumePlugin
14:55:19 
14:55:19 [d539f05da0b31] waiting for daemon to start
14:55:19 [d539f05da0b31] daemon started
14:55:19 
14:55:19 docker_cli_swarm_unix_test.go:22:
14:55:19     // Make sure task stays pending before plugin is available
14:55:19     waitAndAssert(c, defaultReconciliationTimeout, d.CheckServiceTasksInState("top", swarm.TaskStatePending, "missing plugin on 1 node"), checker.Equals, 1)
14:55:19 docker_utils_test.go:465:
14:55:19     c.Assert(v, checker, args...)
14:55:19 ... obtained int = 0
14:55:19 ... expected int = 1
14:55:19 
14:55:19 [d539f05da0b31] exiting daemon
14:55:20 
```

**Comment by RenaudWasTaken** [comment]
After some work I found that the root cause of this bug was introduced by https://github.com/docker/swarmkit/pull/2287

The error message is now surfaced in the Error field. I've added a commit to fix this,

**Comment by thaJeztah** [comment]
Thanks! There's another PR as well that bumps swarm kit, so one of both will have to rebase probably. I created a quick overview of changes included in this PR (currently);

https://github.com/docker/swarmkit/compare/872861d2ae46958af7ead1d5fffb092c73afbaf0...de950a7ed842c7b7e47e9451cde9bf8f96031894

From that list; included in this bump are:

- api: Use TaskStatus Err field for non-terminal errors https://github.com/docker/swarmkit/pull/2287
- Support the init flag for container spec https://github.com/docker/swarmkit/pull/2386
- [Windows] Support for isolation mode on container spec https://github.com/docker/swarmkit/pull/2342 (related to https://github.com/moby/moby/pull/34424)
- Increase gRPC request timeout to 20 seconds for sending snapshots https://github.com/docker/swarmkit/pull/2391
- Delete node attachments when node is removed https://github.com/docker/swarmkit/pull/2409
- add pid limits https://github.com/docker/swarmkit/pull/2415 (related to https://github.com/moby/moby/issues/28618)
- Generic resource cli https://github.com/docker/swarmkit/pull/2347

File: daemon/config/config.go [hunk 1]
```diff
@@ -171,7 +171,8 @@ type CommonConfig struct {
 	Experimental bool `json:"experimental"` // Experimental indicates whether experimental features should be exposed or not
 
 	// Exposed node Generic Resources
-	NodeGenericResources string `json:"node-generic-resources,omitempty"`
+	// e.g: ["orange=red", "orange=green", "orange=blue", "apple=3"]
+	NodeGenericResources []string `json:"node-generic-resources,omitempty"`
```
Comment (thaJeztah) [inline daemon/config/config.go]: Do we have a test to test loading this configuration from a `daemon.json` configuration file?

Comment (RenaudWasTaken) [reply daemon/config/config.go]: Just added one :)

**Comment by anshulpundir** [comment]
Should we also include https://github.com/docker/swarmkit/commit/312be5983e2d2f01f0cbcc99d027b28433256705 ?@thaJeztah

**Review by anshulpundir — COMMENTED** [review]
Should we also include https://github.com/docker/swarmkit/commit/312be5983e2d2f01f0cbcc99d027b28433256705? @thaJeztah

**Comment by RenaudWasTaken** [comment]
Thanks @thaJeztah for the list!
Including it and https://github.com/docker/swarmkit/pull/2419 in the first post.

**Comment by thaJeztah** [comment]
@anshulpundir ah! yes, missed that one in my overview; I updated the list, and updated the list at the top as well

@RenaudWasTaken https://github.com/moby/moby/pull/34424 was merged, so this PR now needs a rebase 😅; probably just amending the first commit ("Revendored swarmkit "), and running `vndr` again should do

**Comment by RenaudWasTaken** [comment]
Looks like it also fixed the integration test: https://github.com/moby/moby/pull/34424/files#diff-eb3f7d0d8dfa41106bb74d7e5dfe43ecR203

I've removed that commit.
Thanks for updating my first post!

**Comment by RenaudWasTaken** [comment]
@thaJeztah is this PR good to go ?
Hoping to revendor the CLI after this so we can get support of GPUs in swarm!

File: opts/opts.go [hunk 2]
```diff
@@ -263,6 +263,16 @@ func ValidateLabel(val string) (string, error) {
 	return val, nil
 }
 
+// ValidateSingleGenericResource validates that a single entry in the
+// generic resource list is valid.
+// i.e 'GPU=UID1' is valid however 'GPU:UID1' or 'UID1' isn't
+func ValidateSingleGenericResource(val string) (string, error) {
+	if strings.Count(val, "=") < 1 {
+		return "", fmt.Errorf("bad attribute format `%s` expected argument of the for `name=value`", val)
```
Comment (thaJeztah) [inline opts/opts.go]: This message looks incomplete ("argument of the for .. ")?

Looking at other messages, I think we use `invalid ...` for other similar messages (e.g. https://github.com/moby/moby/blob/d91c5f42eb37c6f88cec4021c10c0a1ded1785c3/daemon/cluster/convert/container.go#L296)

Perhaps something like:

```
invalid node-generic-resource format: expected 'name=value':
```

@mistyhacks perhaps you have a better suggestion?

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks! Just some minor nits, but looks good otherwise

File: opts/opts.go (hunk 2 above)
Comment (mistyhacks) [reply opts/opts.go]: I think it's ok, if you also include a `got` portion to the error message.  What is a "node-generic-resource"?

Comment (RenaudWasTaken) [reply opts/opts.go]: @mistyhacks the full design document [is here](https://github.com/docker/swarmkit/blob/master/design/generic_resources.md).
The basic idea is that it's a way for a node to advertise user-defined resources

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Review by vdemeester — APPROVED** [review]
LGTM 🐮

**Comment by thaJeztah** [comment]
Thanks @RenaudWasTaken - I think this needs an update to the documentation (as the format for the `dockerd` options changed); unfortunately those are not in this repository, but could you open a pull request for the `dockerd` command-line reference, and the man-page in the https://github.com/docker/cli repository?

- https://github.com/docker/cli/blob/master/docs/reference/commandline/dockerd.md
- https://github.com/docker/cli/blob/master/man/dockerd.8.md

**Comment by RenaudWasTaken** [comment]
@thaJeztah will be doing this as part of docker/cli/pull/429

**Comment by thaJeztah** [comment]
Perfect, thanks @RenaudWasTaken

**Comment by RenaudWasTaken** [comment]
Looks like I'll have to revendor ~100 commits of docker/docker though :(

**Comment by thaJeztah** [comment]
If you prefer, I can do a separate bump for all changes before yours

**Comment by RenaudWasTaken** [comment]
@thaJeztah It would make things a lot easier :) Thanks!

On the other hand it does look like I'll still have to revendor docker  because of the way [parsing is done](https://github.com/moby/moby/blob/f4d4f5863156b82ef146b6ff1e845f8dcf019f12/daemon/config/opts.go#L10).

Basically parsing in the daemon is through swarmkit/api/genericresource.Parse and then docker/daemon/cluster/convert.GenericResourcesFromGRPC

Meaning that if I want to do the same in the CLI I'll need to add a lot of new dependencies. Because that'll mean vendoring docker/daemon/cluster/convert which has a lot of dependencies.

Should I copy swarmkit/api/genericresource.Parse in docker/cli or add the 10+ dependencies to the vendor.conf ?

**Comment by dnephin** [comment]
The dependencies of `swarmkit/api/genericresource/` are fine, it's just `swarmkit/api` which is already a dependency of `docker/cli`.

I would duplicate `convert.GenericResourcesFromGRPC()`. There's not much logic there, and `daemon/cluster/convert` is not an appropriate dependency for `docker/cli`.

**Pull request merged**